File: 01/js/BooleanExpression.js
/*
 * BooleanExpression.js
 *
 * @author Ronald Rey Lovera
 * @contact reyronald@gmail.com
 * @date October, 2015
 */

var PrimeImplicant = require('./PrimeImplicant.js');

var BooleanExpression = (function() {

	BooleanExpression.solveQuineMcCluskey 	= solveQuineMcCluskey;
	BooleanExpression.areEquivalent 		= areEquivalent;

	BooleanExpression.prototype.toString = function(first_argument) {
		return this.reducedExpression;
	};

	return BooleanExpression;

	//////////////////////////////////

	// `Static public` functions
	function solveQuineMcCluskey(expression)
	{
		return new BooleanExpression(expression).toString();
	}

	function areEquivalent(expr1, expr2)
	{
		var terms1 = expr1.split('+').sort();
		var terms2 = expr2.split('+').sort();

		if ( terms1.length !== terms2.length )
			return false;

		for (var t in terms1) 
			terms1[t] = terms1[t].split('').sort().join('');

		for (var t in terms2) 
			terms2[t] = terms2[t].split('').sort().join('');

		return terms1.join('') === terms2.join('');
	}

	// Constructor

	function BooleanExpression(expression)
	{
		this.expression = expression;
		this.variables 	= getVariables(expression);

		// Finding minterms
		this.minterms = getMinterms(this.variables, this.expression);

		// Finding the prime implicants
		this.primeImplicants = getPrimeImplicants(this.minterms, this.variables);

		// Building the prime implicants chart
		this.chart = getPrimeImplicantsChart(this.primeImplicants);

		// Getting a simplified list of necessasry Prime Implicants from the chart
		this.combinedPrimeImplicants = combinePrimeImplicants(this.chart);

		// Gettin the reduced expression string from the combined implicants
		this.reducedExpression = reduceExpression(this.combinedPrimeImplicants, this.variables);
	}

	// `Private` functions

	function getVariables(expression)
	{
		var matches = expression.match(/[A-z]/g).sort();
		return unique(matches);
	}

	function getMinterms(variables, expression)
	{
		var NEGATIVES_REGEX 		= /[A-z]'/g;
		var NEGATIVES_PLAIN_REGEX 	= /([A-z])(?=['])/g;

		function getNegatives(term)
		{
			return term.match(NEGATIVES_PLAIN_REGEX) ? term.match(NEGATIVES_PLAIN_REGEX) : [];
		}

		function getPositives(term)
		{
			return term.replace(NEGATIVES_REGEX, "");
		}

		function getDontCares(term, vars)
		{
			return vars.join("").replace(new RegExp("['" + term + "]" , "g"), "").split("");
		}

		function getTermNumerical(term, variables) 
		{
			var number = 0;
			for (var i in term) {
				var index = variables.indexOf( term[i] );
				if ( index !== -1 )
					number |= 1 << (variables.length - 1 - index );
			}
			return number;
		}

		var terms = expression.split("+");
		var minterms = [];
		for (var t in terms) {

			var positives 	= getPositives(terms[t]);
			var negatives 	= getNegatives(terms[t]);
			var dontCares 	= getDontCares(terms[t], variables);

			var minterm = getTermNumerical(positives, variables);

			if ( minterms.indexOf( minterm ) === -1 )
				minterms.push( minterm );

			// If the term:string we are evaluating in the current 
			// loop is missing some variables, we have to fill
			// the `don't care` bits with all possibilities.
			// e.g. A+ABC
			// A 	=> [ 100, 101, 110, 111 ] (this is the case we're fixing here)
			// ABC 	=> [ 111 ]
			var dC 	= getTermNumerical(dontCares, variables);
			if ( positives.length + negatives.length !== variables.length)
				for (var i = minterm + 1; i < Math.pow(2, variables.length); i++)
					if ( (((dC ^ i) & ~dC) === minterm) && minterms.indexOf( i ) === -1 ) 
						minterms.push( i );
		}
		return minterms;
	}

	function getPrimeImplicants(minterms, variables)
	{
		function countOnes(minterm)
		{
			for (var count = 0; minterm; count++) 
				minterm &= minterm - 1;
			return count;
		}

		function differ(a, b)
		{
			// If `a` and `b` differ only by one bit, return a binary number
			// with a 1 in that position, return false otherwise.
			// e.g. ( 0b100, 0b110 ) => 0b010
			// e.g. ( 0b100, 0b010 ) => false
			return (a&b) === a || (a&b) === b ? a^b : false;
		}

		var table = {};
		for (var i in minterms) {
			var ones = countOnes( minterms[i] );

			if ( typeof table[ones] === 'undefined' )
				table[ ones ] = [];

			table[ ones ].push(new PrimeImplicant({
				value 		: minterms[i],
				dontCares 	: 0,
				marked 		: false,
				finalized 	: false,
				minterms 	: [minterms[i]],
			}));
		}

		var primeImplicants = [];
		var nextTable = {};
		do {
			var completed = true;
			for (var i = 0; i <= variables.length; i++) {
				if (typeof table[i] === 'undefined') continue;

				for (var j = 0; j < table[i].length; j++) {
					//
					if ( typeof table[i+1] === 'undefined'  ) {
						if ( !table[i][j].marked ) {
							table[i][j].finalized = true;
							primeImplicants.push ( table[i][j] );
						}
						continue;
					}

					// Comparing with next level
					var isFinal = true;
					for (var k = 0; k < table[i+1].length; k++) {
						// Only match those with equal `don't care`'s numbers
						if ( table[i][j].dontCares !== table[i+1][k].dontCares ) continue;

						var d = differ( table[i][j].value, table[i+1][k].value );
						if (d !== false) { 
							// Is not final if we found a `matching` prime implicant
							// in the lower level of the table
							isFinal = completed = false;

							table[i][j].marked = table[i+1][k].marked = true;

							if (typeof nextTable[i] === 'undefined')
								nextTable[i] = [];

							nextTable[i].push(new PrimeImplicant({
								value 		: table[i][j].value,
								dontCares 	: table[i][j].dontCares | d,
								marked 		: false,
								finalized 	: false,
								minterms 	: table[i][j].minterms.concat( table[i+1][k].minterms ),
							}));
						}
					}
					if (isFinal && !table[i][j].marked) 
						primeImplicants.push( table[i][j] );
					

				}
			}

			table = nextTable;
			nextTable = {};
		} while ( !completed );

		return PrimeImplicant.reduce( primeImplicants );		
	}

	function getPrimeImplicantsChart(primeImplicants)
	{
		var chart = {};
		for (var i = 0; i < primeImplicants.length; i++) 
			for (var j in primeImplicants[i].minterms) {
				if ( typeof chart[ primeImplicants[i].minterms[j] ] === 'undefined' ) 
					chart[ primeImplicants[i].minterms[j] ] = [];
				chart[ primeImplicants[i].minterms[j] ].push( primeImplicants[i] );
			}
		return chart;		
	}

	function combinePrimeImplicants(chart)
	{
		// TODO: Implement here Petrick's method
		var primeImplicants = [];
		var takenImplicants = [];
		var nonEssentialTaken = false;
		for (var i in chart) {

			for (var j in chart[i]) {
				if ( takenImplicants.indexOf( chart[i][j].toString() ) !== -1 ) continue;

				// Take it if it's an essential implicant.
				// If it's a non-essential implicant, only take it
				// if we haven't picked one before.
				if (!nonEssentialTaken || chart[i].length === 1) {
					primeImplicants.push( chart[i][j] );
					takenImplicants.push( chart[i][j].toString() );
				}

				if (!nonEssentialTaken && chart[i].length > 1) nonEssentialTaken = true;

			}
		}
		return primeImplicants;
	}

	function reduceExpression(primeImplicants, variables)
	{
		// This function takes an array of Prime Implicants,
		// and converts it to a boolean expression in 
		// string form, 
		// e.g. PrimeImplicant[] => `AB'C+B'C'`
		var terms = [];
		for (var i in primeImplicants) {
			var term = [];
			for (var n = 0; n < variables.length; n++) {
				if ( primeImplicants[i].value & primeImplicants[i].dontCares & 1 || !(primeImplicants[i].dontCares & 1) ) 
					term.push(variables[variables.length - 1 - n] + ((primeImplicants[i].value & 1) ? "" : "'" ));
				primeImplicants[i].value >>= 1; 
				primeImplicants[i].dontCares >>= 1;
			}
			terms.push( term.reverse().join("") );
		}
		return unique(terms).sort().join("+");		
	}

	// Misc.

	function unique(array)
	{
		return array.filter(function(elem, pos) {
			return array.indexOf(elem) === pos;
		});
	}	

})();

module.exports = BooleanExpression;

// References
// https://sites.google.com/site/simpogical/download
// https://docs.google.com/viewer?a=v&pid=sites&srcid=ZGVmYXVsdGRvbWFpbnxzaW1wb2dpY2FsfGd4OjZmMTBkNGI4MmRlNDE5MGI
// http://www.ocoudert.com/papers/pdf/int94.pdf
// http://www.ee.surrey.ac.uk/Projects/Labview/minimisation/tabular.html
// https://en.wikipedia.org/wiki/Quine%E2%80%93McCluskey_algorithm
// https://en.wikipedia.org/wiki/Petrick%27s_method
// http://www.quinemccluskey.com/
// http://tma.main.jp/logic/logic.php?lang=en&type=eq&eq=~A~B~C%2B~AB~C%2BA~B~C%2B~ABC%2BA~BC%2BABC
// http://arxiv.org/ftp/arxiv/papers/1404/1404.3349.pdf
// https://webdocs.cs.ualberta.ca/~amaral/courses/329/webslides/Topic5-QuineMcCluskey/sld080.htm

File: 01/dot_net/QuineMcCluskey/QuineMcCluskey/BooleanExpression.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Text.RegularExpressions;
using System.Threading.Tasks;

namespace QuineMcCluskey
{
    public class BooleanExpression
    {
        private readonly string NEGATIVES_PATTERN = "[A-z]'";
        private readonly string NEGATIVES_PLAIN_PATTERN = "([A-z])(?=['])";

        public string Expression { get; set; }
        public char[] Variables { get; set; }
        public IList<int> Minterms { get; set; }
        public IList<PrimeImplicant> PrimeImplicants { get; set; }

        // TODO Improvement: Maybe define a `PrimeImplicantChart` class with an indexer,
        // instead of having this convoluted difficult-to-type Dictionary.
        public IDictionary<int, IList<PrimeImplicant>> Chart { get; set; }

        public IList<PrimeImplicant> SimplifiedPrimeImplicants { get; set; }
        public string ReducedExpression { get; set; }

        public static string SolveQuineMcCluskey(string expression)
        {
            return new BooleanExpression(expression).ToString();
        }

        public static bool AreEquivalent(string expr1, string expr2)
        {
            var terms1 = expr1.ToString().Split('+');
            var terms2 = expr2.ToString().Split('+');

            if (terms1.Length != terms2.Length) return false;

            Array.Sort(terms1);
            Array.Sort(terms2);

            for (int i = 0; i < terms1.Length; i++)
            {
                var chars = terms1[i].ToCharArray();
                Array.Sort(chars);
                terms1[i] = String.Join("", chars);
            }

            for (int i = 0; i < terms2.Length; i++)
            {
                var chars = terms2[i].ToCharArray();
                Array.Sort(chars);
                terms2[i] = String.Join("", chars);
            }

            return String.Join("", terms1) == String.Join("", terms2);
        }
        
        public BooleanExpression(string expression)
        {
            this.Expression = expression;

            this.Variables = this.GetVariables();

            // Finding minterms
            this.Minterms = this.GetMinterms();

            // Finding the prime implicants
            this.PrimeImplicants = this.GetPrimeImplicants();

            // Building the prime implicant chart
            this.Chart = this.GetPrimeImplicantChart();

            // Getting a simplified list of necessasry Prime Implicants from the chart
            this.SimplifiedPrimeImplicants = this.SimplifyPrimeImplicants();

            // Gettin the reduced expression string from the combined implicants
            this.ReducedExpression = this.ReduceExpression();
        }

        public override string ToString()
        {
            return this.ReducedExpression;
        }

        private char[] GetVariables()
        {
            return String.Join("", Regex.Matches(this.Expression, @"[A-z]").OfType<Match>().Select(m => m.Value).Distinct().ToList()).ToCharArray();
        }

        private char[] GetNegatives(string term)
        {
            return String.Join("", Regex.Matches(term, NEGATIVES_PLAIN_PATTERN).OfType<Match>().Select(m => m.Value).Distinct().ToList()).ToCharArray();
        }

        private char[] GetPositives(string term)
        {
            return Regex.Replace(term, NEGATIVES_PATTERN, "").ToCharArray();
        }

        private char[] GetDontCares(string term)
        {
            return Regex.Replace(String.Join("", this.Variables), "[" + term + "]", "").ToCharArray();
        }

        private int GetTermNumerical(char[] positives)
        {
            var number = 0;
            foreach (var variable in positives)
            {
                number |= 1 << (this.Variables.Length - 1 - Array.IndexOf(this.Variables, variable));
            }
            return number;
        }

        private IList<int> GetMinterms()
        {
            var terms = this.Expression.Split('+');

            var minterms = new List<int>();
            foreach (var term in terms)
            {
                var positives = GetPositives(term); 
                var negatives = GetNegatives(term); 
                var dontCares = GetDontCares(term);
                var minterm = GetTermNumerical(positives);

                if (minterms.IndexOf(minterm) == -1)
                    minterms.Add(minterm);

                // If the term:string we are evaluating in the current 
                // loop is missing some variables, we have to fill
                // the `don't care` bits with all possibilities.
                // e.g. A+ABC
                // A 	=> [ 100, 101, 110, 111 ] (this is the case we're fixing here)
                // ABC 	=> [ 111 ]
                // TODO maybe optimize this with a Set that gets smaller in each iteration?
                var dC = GetTermNumerical(dontCares);
                if ( positives.Length + negatives.Length != this.Variables.Length )
                    for (int i = minterm + 1; i < Math.Pow(2, this.Variables.Length); i++)
                    {
                        if ((((dC ^ i) & ~dC) == minterm) && minterms.IndexOf(i) == -1)
                            minterms.Add(i);
                    }

            }
            return minterms;
        }

        private int CountOnes(int minterm)
        {
            int count = 0;
            for (; minterm != 0; count++)
            {
                minterm &= minterm - 1;
            }
            return count;
        }

        private int DifferAt(int a, int b)
        {
            // If `a` and `b` differ only by one bit, return a binary number
            // with a 1 in that position, return -1 otherwise.
            // e.g. ( 0b100, 0b110 ) => 0b010
            // e.g. ( 0b100, 0b010 ) => -1
            return (a & b) == a || (a & b) == b ? a ^ b : -1;
        } 

        private IList<PrimeImplicant> GetPrimeImplicants()
        {
            var table = new Dictionary<int, List<PrimeImplicant>>();

            foreach (var minterm in this.Minterms)
            {
                var ones = CountOnes(minterm);

                if (!table.ContainsKey(ones))
                    table[ones] = new List<PrimeImplicant>();

                table[ones].Add(new PrimeImplicant
                {
                    Value = minterm,
                    DontCares = 0,
                    IsMarked = false,
                    IsFinal = false,
                    Minterms = new List<int>() { minterm }
                });
            }

            var primeImplicants = new List<PrimeImplicant>();
            var nextTable = new Dictionary<int, List<PrimeImplicant>>();
            var isCompleted = true;
            do
            {
                isCompleted = true;
                for (var i = 0; i <= this.Variables.Length; i++)
                {
                    if (!table.ContainsKey(i)) continue;

                    for (var j = 0; j < table[i].Count; j++)
                    {
                        if (!table.ContainsKey(i + 1))
                        {
                            if (!table[i][j].IsMarked)
                            {
                                table[i][j].IsFinal = true;
                                primeImplicants.Add(table[i][j]);
                            }
                            continue;
                        }

                        // Comparing with next level of the table
                        var isFinal = true;
                        for (var k = 0; k < table[i + 1].Count; k++)
                        {
                            // Only match those with equal `don't care`'s numbers
                            if (table[i][j].DontCares != table[i + 1][k].DontCares) continue;

                            var dontCares = DifferAt(table[i][j].Value, table[i + 1][k].Value);
                            if (dontCares != -1)
                            {
                                // Is not final if we found a `matching` prime implicant
                                // in the lower level of the table
                                isFinal = isCompleted = false;

                                table[i][j].IsMarked = table[i + 1][k].IsMarked = true;

                                if (!nextTable.ContainsKey(i))
                                    nextTable[i] = new List<PrimeImplicant>();

                                nextTable[i].Add(new PrimeImplicant
                                {
                                    Value = table[i][j].Value,
                                    DontCares = table[i][j].DontCares | dontCares,
                                    IsMarked = false,
                                    IsFinal = false,
                                    Minterms = table[i][j].Minterms.Concat(table[i + 1][k].Minterms).ToList()
                                });
                            }
                        }
                        if (isFinal && !table[i][j].IsMarked)
                            primeImplicants.Add(table[i][j]);
                    }
                }
                // Resetting for next iteration
                table = nextTable;
                nextTable = new Dictionary<int, List<PrimeImplicant>>();
            } while (!isCompleted);

            return PrimeImplicant.Reduce(primeImplicants);
        }

        private IDictionary<int, IList<PrimeImplicant>> GetPrimeImplicantChart()
        {
            var chart = new Dictionary<int, IList<PrimeImplicant>>();

            foreach (var implicant in this.PrimeImplicants)
            {
                foreach (var minterm in implicant.Minterms)
                {
                    if ( !chart.ContainsKey( minterm ) )
                    {
                        chart[minterm] = new List<PrimeImplicant>();
                    }
                    chart[minterm].Add(implicant);
                }
            }
            return chart;
        }

        private IList<PrimeImplicant> SimplifyPrimeImplicants()
        {
            // TODO: Implement here Petrick's method
            var hasTakenNonEssential = false;
            var implicants = new List<PrimeImplicant>();
            var takenImplicants = new List<string>();

            var keys = this.Chart.Keys.ToList();
            keys.Sort();

            //foreach (KeyValuePair<int, IList<PrimeImplicant>> minterms in this.Chart)
            foreach (var key in keys)
            {
                var minterms = this.Chart[key];
                foreach (var implicant in minterms)
                {
                    if (takenImplicants.Contains(implicant.ToString())) continue;

                    // Take it if it's an essential implicant.
                    // If it's a non-essential implicant, only take it
                    // if we haven't picked one before.
                    if (!hasTakenNonEssential || minterms.Count == 1)
                    {
                        implicants.Add(implicant);
                        takenImplicants.Add(implicant.ToString());
                    }

                    if (!hasTakenNonEssential && minterms.Count > 1) hasTakenNonEssential = true;
                    
                }
            }
            return implicants;
        }

        private string ReduceExpression()
        {
            // This function takes an array of Prime Implicants,
            // and converts it to a boolean expression in 
            // string form, 
            // e.g. PrimeImplicant[] => `AB'C+B'C'`
            var terms = new List<string>();
            var i = 0;
            foreach (var implicant in this.SimplifiedPrimeImplicants) 
            {
                var term = new List<string>();
                for (int n = 0; n < this.Variables.Length; n++)
                {
                    if ( (implicant.Value & implicant.DontCares & 1) != 0 || (implicant.DontCares & 1) == 0)
                    {
                        var quote = (implicant.Value & 1) != 0 ? "" : "'";
                        term.Add(this.Variables[this.Variables.Length - 1 - n] + quote);
                    }
                    this.SimplifiedPrimeImplicants[i].Value >>= 1;
                    this.SimplifiedPrimeImplicants[i].DontCares >>= 1;
                }
                terms.Add( String.Join("", term.Reverse<string>()) );
                i++;
            }
            Array.Sort(terms.Distinct<string>().ToArray());
            return String.Join("+", terms);
        }

    }
}
